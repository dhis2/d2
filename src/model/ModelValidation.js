import { checkType } from '../lib/check';
import Logger from '../logger/Logger';
import Api from '../api/Api';
import { getOwnedPropertyJSON } from './helpers/json';

function extractValidationViolations(webmessage) {
    if (webmessage.response && webmessage.response.errorReports) {
        return webmessage.response.errorReports;
    }

    throw new Error('Response was not a WebMessage with the expected format');
}

/**
 * Handles validation of Model objects based on their modelDefinition.
 *
 * @memberof module:model
 */
class ModelValidation {
    constructor(providedLogger) {
        checkType(providedLogger, 'object', 'logger (Logger)');
        this.logger = providedLogger;
    }

    /**
     * @deprecated Client side model validation is deprecated in favour of server side validation only.
     *
     * @returns {{status: boolean, messages: Array}} Returns {status: true, messages: []}
     */
    validate() {
        this.logger.warn('Client side model validation is deprecated');
        throw new Error('Client side model validation is deprecated');
    }

    /**
     * Sends a POST request against the `api/schemas` endpoint to check if the model is valid.
     *
     * @param {Model} model The model that should be validated.
     * @returns {Array} Returns an array with validation messages if there are any.
     *
     * @note {warn} Currently only checks
     */
    validateAgainstSchema(model) { // eslint-disable-line class-methods-use-this
        if (!(model && model.modelDefinition && model.modelDefinition.name)) {
            return Promise.reject('model.modelDefinition.name can not be found');
        }

        const url = `schemas/${model.modelDefinition.name}`;

        // TODO: The function getOwnedPropertyJSON should probably not be exposed, perhaps we could have a getJSONForModel(ownedPropertiesOnly=true) method.
        return Api.getApi().post(url, getOwnedPropertyJSON(model))
            .then((webMessage) => {
                if (webMessage.status === 'OK') {
                    return [];
                }
                return Promise.reject(webMessage);
            })
            .catch(extractValidationViolations);
    }

    /**
     * Returns the `ModelValidation` singleton. Creates a new one if it does not yet exist.
     * Grabs a logger instance by calling `Logger.getLogger`
     *
     * @returns {ModelValidation} New or memoized instance of `ModelInstance`
     */
    static getModelValidation() {
        if (this.modelValidation) {
            return this.modelValidation;
        }
        return (this.modelValidation = new ModelValidation(Logger.getLogger(console)));
    }
}

export default ModelValidation;
