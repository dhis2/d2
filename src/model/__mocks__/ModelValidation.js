const validation = jest.fn()

export default {
    getModelValidation() {
        return {
            validateAgainstSchema: validation,
        }
    },
}
