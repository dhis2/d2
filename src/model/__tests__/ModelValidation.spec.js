import MockLogger from '../../logger/Logger'
import MockApi from '../../api/Api'
import ModelValidation from '../ModelValidation'

jest.mock(
    '../../logger/Logger',
    () =>
        class Logger {
            static getLogger() {
                return new Logger()
            }
        }
)
jest.mock('../../api/Api')
jest.mock('../helpers/json', () => ({
    getOwnedPropertyJSON() {
        return { id: 'R4dd3wwdwdw', name: 'ANC' }
    },
}))

describe('ModelValidations', () => {
    let mockApi
    let modelValidation

    beforeEach(() => {
        mockApi = MockApi.getApi()

        modelValidation = new ModelValidation(new MockLogger({}))
    })

    afterEach(() => {
        MockApi.mockReset()
    })

    it('should create a ModelValidation object', () => {
        expect(modelValidation).toBeInstanceOf(ModelValidation)
    })

    it('should not be allowed to be called without new', () => {
        expect(() => ModelValidation()).toThrowErrorMatchingSnapshot()
    })

    describe('getModelValidation', () => {
        it('should return a ModelValidation object', () => {
            expect(ModelValidation.getModelValidation()).toBeInstanceOf(
                ModelValidation
            )
        })

        it('should create a singleton and return that', () => {
            expect(ModelValidation.getModelValidation()).toBe(
                ModelValidation.getModelValidation()
            )
        })
    })

    describe('validateAgainstSchema', () => {
        let modelMock

        beforeEach(() => {
            modelMock = {
                modelDefinition: {
                    name: 'dataElement',
                    getOwnedPropertyJSON: jest
                        .fn()
                        .mockReturnValueOnce({
                            id: 'R4dd3wwdwdw',
                            name: 'ANC',
                        }),
                    getOwnedPropertyNames: jest
                        .fn()
                        .mockReturnValueOnce(['id', 'name']),
                    modelValidations: {
                        id: {},
                        name: {},
                    },
                },
                dataValues: { id: 'R4dd3wwdwdw', name: 'ANC' },
                getCollectionChildrenPropertyNames: jest
                    .fn()
                    .mockReturnValueOnce([]),
                getReferenceProperties: jest.fn().mockReturnValueOnce([]),
            }
        })

        it('should be a function', () => {
            expect(modelValidation.validateAgainstSchema).toBeInstanceOf(
                Function
            )
        })

        it('should return a rejected promise if the model.modelDefinition.name is not present', () => {
            modelValidation.validateAgainstSchema().catch(message => {
                expect(message).toBe(
                    'model.modelDefinition.name can not be found'
                )
            })
        })

        it('should call the post method on the Api', () => {
            mockApi.post.mockReturnValueOnce(
                Promise.resolve({
                    httpStatus: 'OK',
                    httpStatusCode: 200,
                    status: 'OK',
                    response: {
                        responseType: 'ValidationViolations',
                    },
                })
            )

            return modelValidation.validateAgainstSchema(modelMock).then(() => {
                expect(mockApi.post).toBeCalled()
            })
        })

        it('should call the post method on the api with the modeldata', () => {
            mockApi.post.mockReturnValueOnce(
                Promise.resolve({
                    httpStatus: 'OK',
                    httpStatusCode: 200,
                    status: 'OK',
                    response: {
                        responseType: 'ValidationViolations',
                    },
                })
            )

            return modelValidation.validateAgainstSchema(modelMock).then(() => {
                expect(mockApi.post).toBeCalledWith('schemas/dataElement', {
                    id: 'R4dd3wwdwdw',
                    name: 'ANC',
                })
            })
        })

        it('should return the validationViolations array from the webmessage', () => {
            const schemaValidationResult = {
                httpStatus: 'Bad Request',
                httpStatusCode: 400,
                status: 'ERROR',
                response: {
                    responseType: 'ErrorReports',
                    errorReports: [
                        {
                            message: 'Missing required property `domainType`.',
                            mainKlass: 'org.hisp.dhis.dataelement.DataElement',
                            errorKlass:
                                'org.hisp.dhis.dataelement.DataElementDomain',
                            errorCode: 'E4000',
                        },
                        {
                            message:
                                'Missing required property `categoryCombo`.',
                            mainKlass: 'org.hisp.dhis.dataelement.DataElement',
                            errorKlass:
                                'org.hisp.dhis.dataelement.DataElementCategoryCombo',
                            errorCode: 'E4000',
                        },
                        {
                            message: 'Missing required property `name`.',
                            mainKlass: 'org.hisp.dhis.dataelement.DataElement',
                            errorKlass: 'java.lang.String',
                            errorCode: 'E4000',
                        },
                        {
                            message: 'Missing required property `shortName`.',
                            mainKlass: 'org.hisp.dhis.dataelement.DataElement',
                            errorKlass: 'java.lang.String',
                            errorCode: 'E4000',
                        },
                    ],
                },
            }
            mockApi.post.mockReturnValueOnce(
                Promise.reject(schemaValidationResult)
            )

            return modelValidation
                .validateAgainstSchema(modelMock)
                .then(validationMessages => {
                    expect(validationMessages).toBe(
                        schemaValidationResult.response.errorReports
                    )
                })
        })

        it('should return the errorReports array from the webmessage', () => {
            const schemaValidationResult = {
                httpStatus: 'Bad Request',
                httpStatusCode: 400,
                status: 'ERROR',
                response: {
                    errorReports: [
                        {
                            message: 'Required property missing.',
                            property: 'name',
                        },
                    ],
                },
            }
            mockApi.post.mockReturnValueOnce(
                Promise.reject(schemaValidationResult)
            )

            return modelValidation
                .validateAgainstSchema(modelMock)
                .then(validationMessages => {
                    expect(validationMessages).toEqual([
                        {
                            message: 'Required property missing.',
                            property: 'name',
                        },
                    ])
                })
        })

        it('should return an empty array when the validation passed', () => {
            mockApi.post.mockReturnValueOnce(
                Promise.resolve({
                    httpStatus: 'OK',
                    httpStatusCode: 200,
                    status: 'OK',
                    response: {
                        responseType: 'ValidationViolations',
                    },
                })
            )

            expect.assertions(1)

            return modelValidation
                .validateAgainstSchema(modelMock)
                .then(validationMessages => {
                    expect(validationMessages).toEqual([])
                })
        })

        it('should throw an error when the server does not return the correct WebMessage format', () => {
            const schemaValidationResult = {
                httpStatus: 'Bad Request',
                httpStatusCode: 400,
                status: 'ERROR',
                response: {},
            }
            mockApi.post.mockReturnValueOnce(
                Promise.reject(schemaValidationResult)
            )

            return modelValidation
                .validateAgainstSchema(modelMock)
                .catch(errorMessage => {
                    expect(errorMessage.message).toBe(
                        'Response was not a WebMessage with the expected format'
                    )
                })
        })

        it(
            'should reject the promise if the server gives a successful status code ' +
                'but the web message status is not the `OK` string',
            () => {
                mockApi.post.mockReturnValueOnce(
                    Promise.resolve({ data: 'someData' })
                )

                return modelValidation
                    .validateAgainstSchema(modelMock)
                    .catch(errorMessage => {
                        expect(errorMessage.message).toBe(
                            'Response was not a WebMessage with the expected format'
                        )
                    })
            }
        )
    })
})
