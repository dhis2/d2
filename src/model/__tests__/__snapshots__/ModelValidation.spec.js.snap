// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelValidations should not be allowed to be called without new 1`] = `"Class constructor ModelValidation cannot be invoked without 'new'"`;
